middle_end/flambda2/to_cmm/to_cmm_env.ml

```diff
@@ -574,47 +495,81 @@ let split_complex_binding ~env ~res (binding : complex binding) =
     let prim_effects =
       Flambda_primitive.Without_args.effects_and_coeffects prim
     in
-    (match To_cmm_effects.classify_by_effects_and_coeffects prim_effects with
-    | Pure | Generative_immutable -> ()
-    | Effect | Coeffect_only ->
-      Misc.fatal_errorf
-        "Primitive %a was marked as `must_inline`, but is has the following \
-         effects and coeffects: %a. This would lead to errors when moving the \
-         primitive application to substitute it."
-        Flambda_primitive.Without_args.print prim Ece.print prim_effects);
+    (* See CR above about effects and coeffets of 'must_inline' primitives *)
+    let effs =
+      match To_cmm_effects.classify_by_effects_and_coeffects prim_effects with
+      | Pure | Generative_immutable -> Ece.pure_can_be_duplicated
+      | Effect | Coeffect_only ->
+        Misc.fatal_errorf
+          "Primitive %a was marked as `must_inline`, but is has the following \
+           effects and coeffects: %a. This would lead to errors when moving \
+           the primitive application to substitute it."
+          Flambda_primitive.Without_args.print prim Ece.print prim_effects
+    in
     let split_binding =
       { order = binding.order;
-        effs = prim_effects;
+        effs;
         inline = binding.inline;
         bound_expr = Split { cmm_expr = new_cmm_expr };
         cmm_var = binding.cmm_var
       }
     in
     res, Split { new_bindings; split_binding }
 
-let remove_binding env var =
-  { env with bindings = Variable.Map.remove var env.bindings }
-
-let will_inline_simple env res { effs; bound_expr = Simple { cmm_expr }; _ } =
-  cmm_expr, env, res, effs
-
-let will_inline_complex env res { effs; bound_expr; _ } =
-  match bound_expr with
-  | Split { cmm_expr } -> cmm_expr, env, res, effs
-  | Splittable_prim { dbg; prim; args } ->
-    let cmm_expr, res = rebuild_prim ~dbg ~env ~res prim (List.map fst args) in
-    cmm_expr, env, res, effs
+(* Adding binding to the env and split them *)
 
-let will_not_inline_simple env res { cmm_var; bound_expr = Simple _; _ } =
-  ( C.var (Backend_var.With_provenance.var cmm_var),
-    env,
-    res,
-    Ece.pure_can_be_duplicated )
+let rec add_binding_to_env ?extra env res var (Binding binding as b) =
+  let env =
+    let bindings = Variable.Map.add var b env.bindings in
+    let cmm_expr = C.var (Backend_var.With_provenance.var binding.cmm_var) in
+    let vars = Variable.Map.add var cmm_expr env.vars in
+    let vars_extra =
+      match extra with
+      | None -> env.vars_extra
+      | Some info -> Variable.Map.add var info env.vars_extra
+    in
+    { env with bindings; vars; vars_extra }
+  in
+  let classification =
+    To_cmm_effects.classify_by_effects_and_coeffects binding.effs
+  in
+  let inline : _ inline = binding.inline in
+  match inline with
+  | Must_inline_and_duplicate -> (
+    match classification with
+    | Pure | Generative_immutable -> env, res
```

Comment:
For `Must_inline_and_duplicate`, it should be possible to always split early, but that doesn't simplify the rest of the code much (we still need to be able to split late for the `Must_inline_once` case). At least documenting that both options are correct would be nice.

Reply:
done